### PR TITLE
[FOR REVIEW, DO NOT MERGE] Parallelized prefetching a list of blocks

### PR DESCRIPTION
**Pull request by ehnuje**
## Proposed changes

- Describe your changes to communicate to the maintainers why we should accept this pull request.
- If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

- This is just a review purpose, DO NOT MERGE THIS
- This PR is made based on https://github.com/klaytn/klaytn/pull/832

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -1632,31 +1638,47 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
...
+				}
+				logger.Info("sent blocks to prefetch channel", "numBlocks", len(chain)-1)
+			}
+		}
 
 		// Process block using the parent state as reference point.
 		receipts, logs, usedGas, internalTxTraces, procStats, err := bc.processor.Process(block, stateDB, bc.vmConfig)
 		if err != nil {
 			bc.reportBlock(block, receipts, err)
+			atomic.StoreUint64(&bc.followupInterrupt, 1)
```
Comment (KimKyungup) [inline blockchain/blockchain.go:1658]: @ehnuje 
This code stop the prefetching future block after every one block.
I think this makes very frequent context switch.
So I suggest to start prefetching all blocks(chain) before `for i, block := range chain {` like `senderCacher()`.
Also we can stop the fetch goroutine after this loop.

I would like you to test this fetch logic also.

**Comment by KimKyungup** [comment]
@aidan-kwon @winnie-byun @jeongkyun-oh 
If you have time, please review this prefetch logic and share your opinion.

@ehnuje For reviews, please explain shortly about this prefetch.

**Comment by aidan-kwon** [comment]
I prefer pre-fetching just one block because of the following reasons. 
1. Prefetcher goroutines usually work in synchronization mode. When a node has the latest state, the goroutines do nothing in the common environment but the goroutines are maintained. I think it wastes node resources. 
2. It is a trade-off between CPU+Memory and DB read. Prefetcher sacrifices CPU and Memory to read DB data faster. However, I cannot sure Klaytn nodes (especially ENs) has enough computation power and memory to process parallel block insert safely. Nodes already have hundreds of goroutines and large size caches.

File: blockchain/blockchain.go (hunk 1 above)
Comment (ehnuje) [reply blockchain/blockchain.go]: As this PR showed worse performance than only fetching next block, I'll just update this PR and compare its performance to the original prefetcher.

Comment (ehnuje) [reply blockchain/blockchain.go]: Please check 21fbc537c66f705af24f5c2b615c378c3c6bebb8

**Comment by ehnuje** [comment]
@aidan-kwon Thanks for your comments. Please check my comments and share your opinion 😄 

> I prefer pre-fetching just one block because of the following reasons.
> 
> 1. Prefetcher goroutines usually work in synchronization mode. When a node has the latest state, the goroutines do nothing in the common environment but the goroutines are maintained. I think it wastes node resources.

Do you know the performance impact or the resource usage of goroutine that does nothing? If it really degrades the performance, maybe we can think about creating and deleting goroutine only if downloader works.

> 2. It is a trade-off between CPU+Memory and DB read. Prefetcher sacrifices CPU and Memory to read DB data faster. However, I cannot sure Klaytn nodes (especially ENs) has enough computation power and memory to process parallel block insert safely. Nodes already have hundreds of goroutines and large size caches.

Regarding EN, I'm not sure how much we should care EN. The block itself is already generated from the high-end instances. Also, as far as I know, Klaytn node does not utilize CPU since block processing is a serial job. If the node serves many requests, using additional CPU might be a problem, but I think using a bit more CPU won't be a problem in most cases.

File: blockchain/state_prefetcher.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+func (p *statePrefetcher) Prefetch(block *types.Block, stateDB *state.StateDB, cfg vm.Config, interrupt *uint64) {
+	var (
+		header = block.Header()
+	)
+	// Iterate over and process the individual transactions
+	for i, tx := range block.Transactions() {
+		// If block precaching was interrupted, abort
+		if interrupt != nil {
+			val := atomic.LoadUint64(interrupt)
+			if val == 1 {
```
Comment (jeongkyun-oh) [inline blockchain/state_prefetcher.go:64]: Isn't it possible to set `interrupt` to 0 in `insertChain` before this method returns?

**Comment by winnie-byun** [comment]
About go routine and resources:
1. The main resource go routine is using is memory usage. (4536bytes are used per routine in go 1.6.2 [link](https://stackoverflow.com/questions/8509152/max-number-of-goroutines#:~:text=On%20a%20machine%20with%204,slightly%20less%20than%201%20million.&text=It%20is%20practical%20to%20create,in%20the%20same%20address%20space.)) 
2. If a go routine is blocked by channel, the routine is only started when there is a signal. (Checkout `Don't interrupt!` in [this article](https://codeburst.io/why-goroutines-are-not-lightweight-threads-7c460c1f155f)) 
3. The uploaded data is stored in state trie cache. If the cache size is in concern, we could resize the cache.
@aidan-kwon @ehnuje

**Pull request closed without merging**